common.py

```diff
@@ -0,0 +1,110 @@
+        # __file__ is only available if executed from a file but not from an ipython shell or notebook
+        # in those cases, the file has to be loaded remotely from github.
+        try:
+            local_path = os.path.abspath(os.path.dirname(__file__))
+            is_local = True
+            if not local_path.endswith('/gender_novels'):
+                is_local = False
+                print(f"WARNING: The FileLoaderMixin should be placed in the main path of the gender_novels project."
+                      f"It's currently in {local_path}. Until the Mixin is in the correct path, files are loaded " \
+                      f"remotely.")
+        except NameError:
+            is_local = False
+
+        if is_local:
+            print(f'loading {file_path} locally.')
+            return self._load_file_locally(file_path)
+        else:
+            print(f'loading {file_path} remotely')
+            return self._load_file_online(file_path)
+
+
+    def _load_file_locally(self, file_path):
+
+        # I need a way of getting the local path to the base of the repo. This file is currently in the base of the
+        # repo so it returns the correct path. But it will change once this function gets moved.
+        local_base_path = os.path.abspath(os.path.dirname(__file__))
+        file = open(f'{local_base_path}/{file_path}', mode='r')
+
+        if file_path.endswith('.csv'):
+            result = file.readlines()
+        elif file_path.endswith('.txt'):
+            result = file.read()
+
+        file.close()
+        return result
+
+    def _load_file_online(self, file_path):
+
+        base_path = 'https://raw.githubusercontent.com/dhmit/gender_novels/master/'
+        url = f'{base_path}/{file_path}'
+        response = urllib.request.urlopen(url)
+
+        if url.endswith('.csv'):
+            return [line.decode('utf8') for line in response.readlines()]
+        elif url.endswith('.txt'):
+            return response.read().decode('utf8')
+
+class Corpus(FileLoaderMixin):
+
+    def __init__(self, corpus_name):
+        self.corpus_name = corpus_name
+        print("here", type(self))
+        self.novels = self._load_novels()
+
+    def _load_novels(self):
+
+        novels = []
+
+        csv_path = f'corpora/{self.corpus_name}/{self.corpus_name}.csv'
+        csv_file = self.load_file(csv_path)
+        csv_reader = csv.DictReader(csv_file)
+
+        for novel_metadata in csv_reader:
+            novel_metadata['corpus_name'] = self.corpus_name
+            novels.append(Novel(novel_metadata_dict=novel_metadata))
+
+        return novels
+
+class Novel(FileLoaderMixin):
+
+    def __init__(self, novel_metadata_dict):
+        for k, v in novel_metadata_dict.items():
+            setattr(self, k, v)
+
+        # Check that the essential attributes for the novel exists.
+        # Currently available attributes that are not checked are: country_publication, author_gender, and notes.
+        assert hasattr(self, 'author')
```

Comment:
I'm not a fan of bare "assert" statements because they raise an error that doesn't tell the end-user what to do to fix it; they have to read the code in order to understand what was wanted of them.  I think that better would be putting something like this at the top at line 91:

```python
if not hasattr(novel_metadata_dict, 'items'):
    raise ValueError('novel_metadata_dict must be a dictionary or support .items()')
for key in ('author', 'date', 'title'):
    if key not in novel_metadata_dict:
        raise ValueError(f'novel_metadata_dict must have an entry for "{key}"')
```

Reply:
Thanks, that's very useful. Added.